numpyro/infer/autoguide.py

```diff
@@ -287,15 +287,15 @@ class AutoDelta(AutoGuide):
     """
     def __init__(self, model, *, prefix='auto', init_loc_fn=init_to_median,
                  create_plates=None):
-        self.init_loc_fn = init_loc_fn
         self._event_dims = {}
         super().__init__(model, prefix=prefix, init_loc_fn=init_loc_fn, create_plates=create_plates)
 
     def _setup_prototype(self, *args, **kwargs):
         super()._setup_prototype(*args, **kwargs)
-        self._init_locs = {
-            k: v for k, v in self._postprocess_fn(self._init_locs).items() if k in self._init_locs
-        }
+        with numpyro.handlers.block():
+            self._init_locs = {
+                k: v for k, v in self._postprocess_fn(self._init_locs).items() if k in self._init_locs
```

Comment:
Looks reasonable. What does `self._postprocess_fn()` do? Should it always be blocked, or only here? I guess an alternative place to block would be in `.__call__()`:
```py
def __call__(self, *args, **kwargs):
    if self.prototype_trace is None:
        with numpyro.handlers.block():
            self._setup_prototype(*args, **kwargs)
```
then you could remove the `block` from `_setup_prototype`...

Reply:
> What does self._postprocess_fn() do? Should it always be blocked, or only here?

Only here I think. `postprocess_fn` transforms unconstrained values to constrained values and also collects deterministic sites.

> I guess an alternative place to block would be in .__call__():

You are right. Do you want to move `block` there? Both places look good to me.

Reply:
Your original design looks good.